File: Samples/Evaluate/how-to-edit-rich-text-in-spreadsheetcontrol-19.1.4-/VB/SpreadsheetRichText/CustomDocumentVisitor.vb
Imports DevExpress.Spreadsheet
Imports DevExpress.XtraRichEdit.API.Native
Public Class CustomDocumentVisitor
    Inherits DocumentVisitorBase
    Private richTextString As RichTextString
    Private endPosition As Integer
    Public ReadOnly Property RichText As RichTextString
        Get
            Return richTextString
        End Get
    End Property
    Public Sub New(ByVal endPos As Integer)
        richTextString = New RichTextString()
        endPosition = endPos
    End Sub
    Public Overrides Sub Visit(ByVal text As DocumentText)
        MyBase.Visit(text)
        Dim runFont As RichTextRunFont = CreateRichTextRun(text.TextProperties)
        richTextString.AddTextRun(text.Text, runFont)
    End Sub
    Public Overrides Sub Visit(ByVal paragraphEnd As DocumentParagraphEnd)
        MyBase.Visit(paragraphEnd)
        If endPosition - 1 <> paragraphEnd.Position Then
            Dim runFont As RichTextRunFont = CreateRichTextRun(paragraphEnd.TextProperties)
            richTextString.AddTextRun(paragraphEnd.Text, runFont)
        End If
    End Sub
    Private Function CreateRichTextRun(ByVal tp As ReadOnlyTextProperties) As RichTextRunFont
        Dim runFont As New RichTextRunFont(tp.FontName, tp.DoubleFontSize \ 2, tp.ForeColor)
        runFont.Bold = tp.FontBold
        runFont.Italic = tp.FontItalic
        runFont.Strikethrough = tp.StrikeoutType = StrikeoutType.Single
        Select Case tp.Script
            Case DevExpress.Office.CharacterFormattingScript.Subscript
                runFont.Script = ScriptType.Subscript
            Case DevExpress.Office.CharacterFormattingScript.Superscript
                runFont.Script = ScriptType.Superscript
            Case Else
                runFont.Script = ScriptType.None
        End Select
        Select Case tp.UnderlineType
            Case DevExpress.XtraRichEdit.API.Native.UnderlineType.Single
                runFont.UnderlineType = DevExpress.Spreadsheet.UnderlineType.Single
            Case DevExpress.XtraRichEdit.API.Native.UnderlineType.Double
                runFont.UnderlineType = DevExpress.Spreadsheet.UnderlineType.Double
            Case Else
                runFont.UnderlineType = DevExpress.Spreadsheet.UnderlineType.None
        End Select
        Return runFont
    End Function
End Class

File: Samples/Evaluate/how-to-edit-rich-text-in-spreadsheetcontrol-19.1.4-/CS/SpreadsheetRichText/CustomDocumentVisitor.cs
using DevExpress.Spreadsheet;
using DevExpress.XtraRichEdit.API.Native;

namespace SpreadsheetRichText
{
    public class CustomDocumentVisitor : DocumentVisitorBase
    {
        private RichTextString richTextString;
        private int endPosition;
        public RichTextString RichText
        {
            get { return richTextString; }
        }
        public CustomDocumentVisitor(int endPos)
        {
            richTextString = new RichTextString();
            endPosition = endPos;
        }
        public override void Visit(DocumentText text)
        {
            base.Visit(text);
            RichTextRunFont runFont = CreateRichTextRun(text.TextProperties);
            richTextString.AddTextRun(text.Text, runFont);
        }
        public override void Visit(DocumentParagraphEnd paragraphEnd)
        {
            base.Visit(paragraphEnd);
            if (endPosition - 1 != paragraphEnd.Position)
            {
                RichTextRunFont runFont = CreateRichTextRun(paragraphEnd.TextProperties);
                richTextString.AddTextRun(paragraphEnd.Text, runFont);
            }
        }
        private RichTextRunFont CreateRichTextRun(ReadOnlyTextProperties tp)
        {
            RichTextRunFont runFont = new RichTextRunFont(tp.FontName, tp.DoubleFontSize / 2, tp.ForeColor);
            runFont.Bold = tp.FontBold;
            runFont.Italic = tp.FontItalic;
            runFont.Strikethrough = tp.StrikeoutType == StrikeoutType.Single;
            switch(tp.Script)
            {
                case DevExpress.Office.CharacterFormattingScript.Subscript:
                    runFont.Script = ScriptType.Subscript;
                    break;
                case DevExpress.Office.CharacterFormattingScript.Superscript:
                    runFont.Script = ScriptType.Superscript;
                    break;
                default:
                    runFont.Script = ScriptType.None;
                    break;

            }
            switch (tp.UnderlineType)
            {
                case DevExpress.XtraRichEdit.API.Native.UnderlineType.Single:
                    runFont.UnderlineType = DevExpress.Spreadsheet.UnderlineType.Single;
                    break;
                case DevExpress.XtraRichEdit.API.Native.UnderlineType.Double:
                    runFont.UnderlineType = DevExpress.Spreadsheet.UnderlineType.Double;
                    break;
                default:
                    runFont.UnderlineType = DevExpress.Spreadsheet.UnderlineType.None;
                    break;
            }
            return runFont;
        }
    }
}
